Lesson2/SecondHomework.playground/Contents.swift

```diff
@@ -0,0 +1,50 @@
+/// An object, which is a list of elements with a type integer, organized according to the LIFO principle.
+class Stack {
+    fileprivate var array: [Int] = []
+    fileprivate var topElement: Int?
+    
+    func push(element: Int) {
+        array.append(element)
+        topElement = element
+    }
+    
+    func pop() -> Int? {
+        if array.isEmpty {
+            return nil
+        } else {
+            let lastElement = array.removeLast()
+            topElement = array.isEmpty ? nil : array[array.count - 1]
+            return lastElement
+        }
+    }
+    
+    func top() -> Int? {
+        return topElement
+    }
+}
+
+/// An object, which is a list of elements with a type integer, organized according to the LIFO principle.
+class StackStatistics: Stack {
+    private var minimalValue: Int?
+    
+    override func push(element: Int) {
```

Comment:
Could you please try to evaluate the performance using big-O notation?

Reply:
Algorithm complexity - O(n)

Reply:
Yes, you're absolutely right!

The issue here is the following:
the expected time complexity for `push`/`pop` operations is `O(1)` (I meant a classical stack).

Your `StackStatistics` should meet these expectations + should add a new method to calculate a minElement in `O(1)`. Nobody wants to have a stack with linear `push`/`pop` operations. Try to avoid performance degradations whenever possible.

Does that make sense?

Reply:
Yes it makes sense. I made so that the complexity of the algorithm for the push / pop / minimumElement functions was equal to O(1)